python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -399,12 +409,116 @@ def get_known_used_data(
             }
         return {}
 
+    def _calculate_partition_data_time(
+        self,
+        asset_key: AssetKey,
+        record_id: int,
+        partitions_def: TimeWindowPartitionsDefinition,
+        status_cache_value: "AssetStatusCacheValue",
+    ) -> Optional[datetime.datetime]:
```

Comment:
I think some documentation of what's going on here at a high level (like included in the PR description) would be helpful:
1. How the data time for a time-partitioned asset is defined.
2. How we use the status_cache_value to compute it